openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```diff
@@ -0,0 +1,172 @@
+    """
+
+    class Meta:
+        proxy = True
+
+    @classmethod
+    def valid_for(cls, taxonomy: Taxonomy = None, **kwars):
+        """
+        Returns True if the the taxonomy is system-defined
+        """
+        return super().valid_for(taxonomy=taxonomy) and taxonomy.system_defined
+    
+class ClosedSystemObjectTag(ClosedObjectTag):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    @classmethod
+    def valid_for(cls, taxonomy: Taxonomy = None, tag: Tag = None, **kargs):
+        """
+        Returns True if the the taxonomy is system-defined
+        """
+        return super().valid_for(taxonomy=taxonomy, tag=tag) and taxonomy.system_defined
+
+
+class ModelObjectTag(OpenSystemObjectTag):
+    """
+    Object tags used with tags that relate to the id of a model
+    
+    This object tag class is not registered as it needs to have an associated model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = None
+
+    @classmethod
+    def valid_for(cls, taxonomy: Taxonomy = None, **kwars):
+        """
+        Validates if has an associated Django model that has an Id
+        """
+        if not super().valid_for(taxonomy=taxonomy) or not cls.tag_class_model:
+            return False
+        
+        # Verify that is a Django model
+        if not issubclass(cls.tag_class_model, models.Model):
+            return False
+
+        # Verify that the model has 'id' field
+        try:
+            cls.tag_class_model._meta.get_field('id')
+        except FieldDoesNotExist:
+            return False
+
+        return True
+    
+    def _check_instance(self):
+        """
+        Validates if the instance exists
+        """
+        try:
+            intance_id = int(self.value)
+        except ValueError:
+            return False
+        return self.tag_class_model.objects.filter(id=intance_id).exists()
+
+    def _check_tag(self):
+        """
+        Validates if the instance exists
+        """
+        if not super()._check_tag():
+            return False
+        
+        # Validates if the instance exists
+        if not self._check_instance():
+            return False
+        
+        return True
+
+
+class UserObjectTag(ModelObjectTag):
+    """
+    Object tags used on taxonomies associated with user model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = get_user_model()
+
+
+class LanguageObjectTag(ClosedSystemObjectTag):
+    """
+    Object tag for Languages
+
+    The tags are filtered and validated taking into account the 
+    languages available in Django LANGUAGES settings var
+    """
+
+    class Meta:
+        proxy = True
+
+    @classmethod
+    def get_tags(cls, taxonomy: Taxonomy) -> List[Tag]:
+        """
+        Returns a list of tags of the available languages.
+        """
+        # TODO we need to overweite this
+        # tags = super().get_tags()
```

Comment:
I'm not sure where's the best place to call this custom class method..

Maybe in `Taxonomy.get_tags`? e.g.,

```
def get_tags(self):
   ObjectTagClass = self.object_tag_class  # might need to catch ImportError..
   if ObjectTagClass and hasattr(ObjectTagClass, 'get_tags'):
       return ObjectTagClass.get_tags(self)

   ...
```

Reply:
@pomegranited I think it would be better if the `ObjectTag` function returns a filtered query set, which is used in `Taxonomy.get_tags` . This is so that `Taxonomy.get_tags` is always consistent in the data it returns.

**LanguageObjectTag.get_tags_query_set**

```
class LanguageObjectTag(ClosedSystemObjectTag):
@classmethod
    def get_tags_query_set(cls, taxonomy: Taxonomy) -> models.QuerySet:
        """
        Returns a query set of available languages tags.
        """
        available_langs = cls._get_available_languages()
        return taxonomy.tag_set.filter(external_id__in=available_langs)
```

**Taxonomy.get_tags**

```
def get_tags(self):
   query_set = self.tag_set
   ObjectTagClass = self.object_tag_class  # might need to catch ImportError..
   if ObjectTagClass and hasattr(ObjectTagClass, 'get_tags_query_set'):
       query_set = ObjectTagClass.get_tags_query_set(self)

   # Use query set in operations here
```